pipeline/sources/webcam_worker.py

```diff
@@ -0,0 +1,225 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import re
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from pathlib import Path
+from subprocess import PIPE, Popen  # noqa: S404
+
+from dagster import AssetExecutionContext, PipesSubprocessClient
+from jinja2 import Environment, PackageLoader, Template, select_autoescape
+
+
+class CommandExecutionFailed(RuntimeError): ...
+
+
+@dataclass
+class WebcamWorkerConfig:
+    host: str
+    user: str
+    password: str
+    image_path: Path
+    symlink_path: Path
+    keep_days: int
+    worker_count: int
+    check_empty_files: bool = True
+    remote_dir: str = '/'
+
+
+@dataclass
+class SymlinkItem:
+    path: Path
+    name: str
+    filename: str
+    moment: datetime
+
+
+class WebcamWorker:
+    config: WebcamWorkerConfig
+    pipes_subprocess_client: PipesSubprocessClient
+    context: AssetExecutionContext
+    index_template: Template
+
+    def __init__(
+        self,
+        config: WebcamWorkerConfig,
+        context: AssetExecutionContext,
+        pipes_subprocess_client: PipesSubprocessClient,
+    ):
+        self.config = config
+        # TODO: how to get stdout from PipesSubprocessClient?
+        self.pipes_subprocess_client = pipes_subprocess_client
+        self.context = context
+
+        jinja2_env = Environment(loader=PackageLoader(package_name='pipeline'), autoescape=select_autoescape())
+        self.index_template = jinja2_env.get_template('webcam_index.html.j2')
+
+    def run(self):
+        # TODO: splitting up in incremental and full download
+        self.download()
+
+        # We always symlink, as we need to publish the latest images
+        self.symlink_with_index()
+
+        # We just clean at full syncs
+        self.clean()
+
+    def download(self):
+        result, _ = self._run_command(
+            command=[
+                'lftp',
+                '-e',
+                f'mirror --newer-than=now-{self.config.keep_days - 1}days -c --parallel={self.config.worker_count} {self.config.remote_dir} {self.config.image_path}; quit;"',
+                '-u',
+                f'{self.config.user},{self.config.password}',
+                self.config.host,
+            ],
+        )
+        match = re.match(r'Total: (\d*) directories?, (\d*) files?, (\d*) symlinks?', result)
+
+        if match is None:
+            self.context.log.error(f'Could not parse lftp output: {result}')
+            return
+
+        directories, files, symlinks = match.groups()
+
+        # TODO: this would be perfect for metrics
+        self.context.log.info(f'Downloaded {files} files, {directories} directories, {symlinks} symlinks')
+
+    def symlink_with_index(self):
```

Comment:
[nitpick]

Shall we
- pull move the `symlink()` call up into the calling function and pass the resulting list of "latest paths" in here,
- rename this function e.g. `generate_index_page()` (to make it clear that it has a side effect)?

Reply:
Done